go-controller/pkg/ovn/controller/apbroute/external_controller.go

```diff
@@ -0,0 +1,239 @@
+package apbroute
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	adminpolicybasedrouteapi "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/adminpolicybasedroute/v1"
+	adminpolicybasedroutelisters "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/adminpolicybasedroute/v1/apis/listers/adminpolicybasedroute/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	ktypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+)
+
+type gatewayInfoList []*gatewayInfo
+
+func (g gatewayInfoList) String() string {
+
+	s := strings.Builder{}
+	for _, item := range g {
+		s.WriteString(fmt.Sprintf("%s, ", item.gws))
+	}
+	return s.String()
+}
+
+func (g gatewayInfoList) HasIP(ip string) bool {
+	for _, i := range g {
+		if i.gws.Has(ip) {
+			return true
+		}
+	}
+	return false
+}
+
+func (g gatewayInfoList) Insert(items ...*gatewayInfo) gatewayInfoList {
+	for _, item := range items {
+		for _, ip := range item.gws.UnsortedList() {
+			if g.HasIP(ip) {
+				klog.Warningf("Duplicated static IP %s", ip)
+				continue
+			}
+			g = append(g, item)
+		}
+	}
+	return g
+}
+func (g gatewayInfoList) Delete(item *gatewayInfo) gatewayInfoList {
+	for index, i := range g {
+		if i.gws.Equal(item.gws) {
+			g = append(g[:index], g[index+1:]...)
+		}
+	}
+	return g
+}
+
+func (g gatewayInfoList) Len() int {
+	return len(g)
+}
+
+func (g gatewayInfoList) Less(i, j int) bool { return lessGWSIP(g[i], g[j]) }
+func (g gatewayInfoList) Swap(i, j int)      { g[i], g[j] = g[j], g[i] }
+
+func lessGWSIP(l, r *gatewayInfo) bool {
+
+	for lip := range l.gws {
+		for rip := range r.gws {
+			if lip > rip {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+type gatewayInfo struct {
+	gws        sets.Set[string]
+	bfdEnabled bool
+}
+
+type ExternalRouteInfo struct {
+	sync.Mutex
+	Deleted bool
+	PodName ktypes.NamespacedName
+	// PodExternalRoutes is a cache keeping the LR routes added to the GRs when
+	// external gateways are used. The first map key is the podIP (src-ip of the route),
+	// the second the GW IP (next hop), and the third the GR name
+	PodExternalRoutes map[string]map[string]string
+}
+
+type routePolicy struct {
+	labelSelector   *metav1.LabelSelector
+	staticGateways  gatewayInfoList
+	dynamicGateways map[ktypes.NamespacedName]*gatewayInfo
+}
+
+type externalPolicyManager struct {
+	stopCh <-chan struct{}
+
+	// route policies
+
+	routeLister adminpolicybasedroutelisters.AdminPolicyBasedExternalRouteLister
+
+	// Pods
+	podLister corev1listers.PodLister
+
+	// Namespaces
+	namespaceLister corev1listers.NamespaceLister
+
+	// cache for set of policies impacting a given namespace
+	namespaceInfoCache      map[string]*namespaceInfo
+	namespaceInfoCacheMutex *sync.Mutex
+	namespaceLockMap        map[string]*sync.Mutex
```

Comment:
rather than this, why not use the custom lockable key cache we have here: https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/syncmap/syncmap.go#L30